src/gui/map/map_editor.cpp

```diff
@@ -3850,42 +3850,41 @@ void MapEditorController::addMapPart()
 void MapEditorController::removeMapPart()
 {
 	auto* part = map->getCurrentPart();
-	auto i = part->getNumObjects();
+	auto num_objects = part->getNumObjects();
 	
-	QMessageBox::StandardButton button =
-	        QMessageBox::question(
-	            window,
+	if (num_objects > 0)
+	{
+		if (QMessageBox::question(
+                window,
                 tr("Remove current part"),
-                i ? tr("Do you want to remove map part \"%1\" and its %2 objects?").arg(part->getName()).arg(i) :
-                    tr("Do you want to remove empty map part \"%1\"?").arg(part->getName()),
-                QMessageBox::Yes | QMessageBox::No );
+                tr("Do you want to remove map part \"%1\" and its %2 objects?").arg(part->getName()).arg(num_objects),
+                QMessageBox::Yes | QMessageBox::No ) == QMessageBox::No)
+		return;
+	}
 	
-	if (button == QMessageBox::Yes)
+	auto index = map->getCurrentPartIndex();
+	UndoStep* undo_step = new MapPartUndoStep(map, MapPartUndoStep::AddMapPart, index);
+	
+	if (num_objects > 0)
 	{
-		auto index = map->getCurrentPartIndex();
-		UndoStep* undo_step = new MapPartUndoStep(map, MapPartUndoStep::AddMapPart, index);
-		
-		if (i > 0)
+		auto* add_step = new AddObjectsUndoStep(map);
+		do
 		{
-			auto* add_step = new AddObjectsUndoStep(map);
-			do
-			{
-				--i;
-				auto* object = part->getObject(i);
-				add_step->addObject(i, object);
-				part->releaseObject(object);
-			}
-			while (i > 0);
-			
-			auto* combined_step = new CombinedUndoStep(map);
-			combined_step->push(add_step);
-			combined_step->push(undo_step);
-			undo_step = combined_step;
+			--num_objects;
+			auto* object = part->getObject(num_objects);
+			add_step->addObject(num_objects, object);
+			part->releaseObject(object);
 		}
+		while (num_objects > 0);
```

Comment:
The loop modifies the `num_objects` variable. While it's fine as is, it may cause trouble in future code maintenance, if the coder does not spot the value change and assumes the original `num_objects` value. I suggest that the value is left intact even after the part removal.

```suggestion
		for (auto i = 0; i < num_objects; ++i)
		{
			auto* object = part->getObject(i);
			add_step->addObject(i, object);
			part->releaseObject(object);
		}
```

Yes, the new code moves the objects into the undo in the reverse order, compared to the original code.

Reply:
Libor, thank you for your thourough review.

I basically agree with your proposal, but for a different reason, since the function itself is so small that I don't see the risk that someone would oversee the modification of the num_objects variable (which I otherwise would make a const one).
But I really like the usage of a for loop instead of a while loop.

However, I was really surprised that moving the objects into the undo in the reverse order is not an arbitrary choice. Instead, Mapper would crash when undoing the map part removal!
As an alternative I propose to use `for (auto i = num_objects - 1; i >= 0; --i)` to preserve the original order.
If you are fine with it, just give me the thumbs up.